internal/status.go

```diff
@@ -0,0 +1,148 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/dustin/go-humanize"
+)
+
+type StatusLine struct {
+	resources []Resource
+
+	resourceSizes          []int64
+	totalBytes             int64
+	numBytesDownloaded     int64
+	numResourcesDownloaded int
+	spinI                  int
+	indexCh                chan int
+	startTime              time.Time
+	sizingErr              error
+	ctx                    context.Context
+}
+
+var spinChars = [5]string{"-", "\\", "|", "/", "-"}
+
+// NewStatusLine creates, initializes, and starts a new StatusLine.
+func NewStatusLine(resources []Resource, ctx context.Context) (*StatusLine, error) {
+	st := StatusLine{}
+	st.resources = resources
+	st.indexCh = make(chan int)
+	st.ctx = ctx
+	st.getResourcesSizes(resources, 10000)
+	st.start()
+	return &st, st.sizingErr
+}
+
+// Increment informs the StatusLine that a resource (at index i in resource list) has finished downloading.
+func (st *StatusLine) Increment(i int) {
+	st.indexCh <- i
+}
+
+// start begins the goroutine and loop that will update/print the status line.
+func (st *StatusLine) start() {
+	st.startTime = time.Now()
+
+	go func() {
+		st.spinI = 0
+
+		for {
+
+			// Block until value is inserted into indexCh (>= 0 when resource finishes downloading, -1 every 50 milliseconds to keep timer and spinner updating).
+			var i int
+			select {
+			case i = <-st.indexCh:
+			case <-time.After(50 * time.Millisecond):
+				i = -1
+			case <-st.ctx.Done():
+				return
+			}
+			if i != -1 {
+				st.numBytesDownloaded += st.resourceSizes[i]
+				st.numResourcesDownloaded++
+			}
+
+			// Update/rotate spinner.
+			st.spinI += 1
+			if st.spinI == len(spinChars) {
+				st.spinI = 0
+			}
+
+			line := st.composeStatusString()
+			fmt.Print(line)
+
+			if st.numResourcesDownloaded == len(st.resources) {
+				fmt.Println()
+				return
+			}
+
+		}
+	}()
+
+}
+
+func (st *StatusLine) getResourcesSizes(resources []Resource, timeoutMs int) {
```

Comment:
This should probably be renamed `initResourcesSizes` now that it doesn't return anything. And it should return an error in case something goes wrong.

Reply:
Right now, if an error occurs, initResourceSizes stores the error in the sizingErr field. This is then returned from NewStatusLine for the caller to handle and also used in composeStatusString to change the output. 

Should this be changed?

Reply:
Okay, that's probably fine. But you should still rename the method since it doesn't return anything.

Reply:
Would it be better for initResourceSizes to return the error (to better match the pattern of functions returning their errors in the rest of the codebase) -- then NewStatusLine could store it in sizingErr (for later use in composeStatusString) and finally return it to the caller?

Reply:
I think it would be preferable, yes. That's the normal pattern that golang developers would expect.